### DISCUSSION TITLE
Initialise MemoryBlock

### DISCUSSION THREAD

**Discussion by dnaldoog** [RomanKubiak/ctrlr#483] [Show and tell]
Here is a method added to the class MemoryBlock() for initialising a MemoryBlock (usually with zeroes).
The overloaded JUCE function `m=MemoryBlock(20)` is not implemented or bound to lua (at least in version 5.3.201), so normally one would write something like `m=MemoryBlock("00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00")` which is a bit cumbersome or use `createFromTable`

local t={}
for i=1,20 do
t[i]=0
end
m:createFromTable(t)

but this following way is maybe a bit more convenient;

### Edit: 03/15/2023

Updated code to support newer versions of Ctrlr.

<pre>
<code>
    function MemoryBlock:initialise(size) -- initialise a Memory Block with data if necessary
        --[[ 
    Versions 5.3.* do not support initialisation of MemoryBlock() with JUCE constructor:
    MemoryBlock (const size_t initialSize, bool initialiseToZero=false)
	This function just offers compatibility between 5.3* and later versions
	    ]]
        local major, minor = utils.getVersionMajor(), utils.getVersionMinor()
        major = tonumber(major)
        minor = tonumber(minor)
        if major > 4 and minor > 3 then
            local tmp = MemoryBlock(size, true) -- use luabind JUCE constructor
            self:loadFromHexString(tmp:toHexString(1))
        else
            local t = {}
            for i = 1, size do
                t[i] = 0
            end
            if self:getSize() > 0 then
                self:removeSection(0, self:getSize())
            end
            self:createFromTable(t)
        end -- looking for 5.3.201/5.3.198
    end
</code>
</pre>

**Comment by dobo365** [comment]
Great explanation John!
Yes indeed, I also figured out long time ago that the initialization of a MemoryBlock based on size was not implemented in 5.3.201.

What I'm doing is thus similar to your approach...
- in a Tables_Init method called at panel load, I pre-define some tables that I create the MB from
- in other methods, if necessary, I use the append() function to extend my initial MB

**Code 1**
`	mbWave128 = MemoryBlock ()
	mbWave128:createFromTable({
	--	00	  01    02    03	04    05	06	  07	08	  09	10	  11	12	  13	14	  15	16    17	18    19
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}) 
`

**Code 2**
`		mbTemp = MemoryBlock()
		for i=1,64 do
			mbTemp:append(mbWave128)
		end
`

Sorry... line returns seems lost...

**Reply by dnaldoog** [reply]
Thanks Goodweather,

Yes your way works perfectly.

With this code, to me what is convenient is that the method `initialize`  extends the `MemoryBlock()` class. Kind of makes it more readable when coming back to it in 6 months' time etc! 😃

Incidentally, you can create line returns using the html tags &lt;pre&gt; &#x2026; &lt;/pre&gt;

**Comment by spiffo-smith** [comment]
dnaldoog, that post was extremely timely, thank you ever so much for that!

I had a MemoryBlock that was normally initialised from a MidiMessage but I needed to initialise it 'on Panel Load' but after lots of reference to the JUCE libraries and trying just about every combination I could think of I just couldn't get anything to work!

Your post has cleared all that up and now everything's working great :-)

**Reply by dnaldoog** [reply]
Great to hear Spiffo!

I was hoping it might be useful for someone!
